aggregator_core/src/datastore.rs

```diff
@@ -1440,26 +1430,27 @@ impl<C: Clock> Transaction<'_, C> {
         let stmt = self
             .prepare_cached(
                 "INSERT INTO client_reports (
-                    task_id,
-                    report_id,
-                    client_timestamp,
-                    extensions,
-                    public_share,
-                    leader_input_share,
-                    helper_encrypted_input_share,
-                    created_at,
-                    updated_at,
+                    task_id, report_id, client_timestamp, extensions, public_share,
+                    leader_input_share, helper_encrypted_input_share, created_at, updated_at,
                     updated_by
                 )
                 VALUES (
                     (SELECT id FROM tasks WHERE task_id = $1), $2, $3, $4, $5, $6, $7, $8, $9, $10
                 )
-                ON CONFLICT DO NOTHING
-                RETURNING COALESCE(client_timestamp < COALESCE($3::TIMESTAMP - (SELECT report_expiry_age FROM tasks WHERE task_id = $1) * '1 second'::INTERVAL, '-infinity'::TIMESTAMP), FALSE) AS is_expired",
+                ON CONFLICT(task_id, report_id) DO UPDATE
+                    SET (
+                        client_timestamp, extensions, public_share, leader_input_share,
+                        helper_encrypted_input_share, created_at, updated_at, updated_by
+                    ) = (
+                        excluded.client_timestamp, excluded.extensions, excluded.public_share,
+                        excluded.leader_input_share, excluded.helper_encrypted_input_share,
+                        excluded.created_at, excluded.updated_at, excluded.updated_by
+                    )
+                    WHERE client_reports.client_timestamp < COALESCE($11::TIMESTAMP - (SELECT report_expiry_age FROM tasks WHERE id = client_reports.task_id) * '1 second'::INTERVAL, '-infinity'::TIMESTAMP)",
```

Comment:
What prevents a client from changing an existing report's input shares, public shares, etc? i.e.:

- report gets uploaded, gets written via put_client_report
- report gets aggregated, scrubbed
- evil client uploads report with same report ID + task ID, but changes input shares

Reply:
Note that, due to the `WHERE` clause on the update, we will only update a row that is already (logically) GC'ed. So the answer, in the datastore, is: nothing, just like with other rows that have been deleted. But there was no such protection before, either -- once the (physical) GC ran, the row would be gone, and nothing in the datastore would the report ID from being reused. The change here is to treat logical GC the same as physical GC.

The protection against this lives in the upload handler, which will reject reports which would be immediately expired:

https://github.com/divviup/janus/blob/5db944a7952c5e49756868cb14a45e9854a72d0e/aggregator/src/aggregator.rs#L1564-L1574

Reply:
I'm uneasy about allowing mutations to columns that were previously write-once as well. Note that there would now be a race condition where a process that started a query with an older value of `clock.now()` may see the `ON CONFLICT DO UPDATE` write from another process.